src/mococrw/ecies.h

```diff
@@ -0,0 +1,400 @@
+     * @brief Sets the ephemeral public key which was generated during encryption
+     *
+     * For decrypting the ciphertext the ephemeral public key, which was generated during encryption, is needed.
+     * @param ephKey The ephemeral key generated during encryption
+     */
+    void setEphemeralKey(const AsymmetricPublicKey& ephKey);
+
+    /**
+     * @brief Set the authentication tag
+     *
+     * As the last step during encryption the authentication tag is calculated. Set this tag to verify the authenticity
+     * of the received message.
+     * @param tag The authentication tag
+     */
+    void setMAC(const std::vector<uint8_t>& tag);
+
+private:
+    friend ECIESCtxBuilder;
+    ECIESDecryptionCtx(const ECIESCtxBuilder &ctxBuilder);
+
+    /**
+     * Internal class for applying the PIMPL design pattern
+     * (to hide the details of storing the padding objects from the client)
+     */
+    class Impl;
+
+    /**
+     * Pointer for PIMPL design pattern
+     */
+    std::unique_ptr<Impl> _impl;
+};
+
+/**
+ * @brief The ECIESCtxBuilder class is an pseudo abstract class for providing common functions to the Encryption and
+ * Decryption context builders.
+ */
+class ECIESCtxBuilder
+{
+public:
+    /**
+     * @brief constructor
+     */
+    ECIESCtxBuilder();
+
+    ~ECIESCtxBuilder();
+
+    /**
+     * @brief Sets the key derivation function used for ECIES
+     * @param kdf The actual object used in ECIES
+     * @return builder instance
+     */
+    ECIESCtxBuilder &setKDF(std::shared_ptr<KeyDerivationFunction> kdf);
+
+    /**
+     * @brief Sets the required key size for MAC ciphers
+     *
+     * Default: The default MAC-Cipher is HMAC-SHA512, thus the key size is 512/8 = 64
+     * @param keySize The key size in bytes
+     * @return builder instance
+     */
+    ECIESCtxBuilder &setMacKeySize(size_t keySize);
+
+    /**
+     * @brief Sets a function for creating the MAC object inside ECIES
+     *
+     * Default: HMAC-SHA512 if not set
+     *
+     * As the key for MAC is generated during ECIES key generation, the MAC object has to be instantiated after that.
+     * This function is a dependency injection.
+     * (example using HMAC-SHA512:
+     \code{.cpp}
+        auto macFunc = [](const std::vector<uint8_t> &key) -> std::unique_ptr<MessageAuthenticationCode> {
+            return std::make_unique<mococrw::HMAC>(openssl::DigestTypes::SHA512, key);
+        };
+     \endcode
+     * )
+     * @param func A std::function which returns an object implementing the MessageAuthenticationCode interface
+     * @return builder instance
+     */
+    ECIESCtxBuilder &setMacFactoryFunction(std::function<std::unique_ptr<MessageAuthenticationCode>(
+                                               const std::vector<uint8_t>&)> func);
+
+    /**
+     * @brief Sets the required key size for symmetric ciphers
+     *
+     * Default: The default symmetric cipher is AES-CBC PKCS7, thus the key length is 256/8 = 32
```

Comment:
```suggestion
     * Default: The default symmetric cipher is AES-CBC PKCS7 with a key length of 256/8 = 32
```
Exists as well in 128bit key size. PKCS#7 is the padding format.

Reply:
Donne